### PR TITLE
Support DNS period resolving mechanism

### LINKED ISSUE #6222: dns resolve for skywalking trace receiver service addresses
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [X ] Question or discussion
- [ ] Bug
- [ ] Requirement
- [ ] Feature or performance improvement

___
### Question
- What do you want to know?
I want to add dns resolver for the domain of  skywalking trace receiver service
-  WHY ?
   - It's hard to maintain agent config
`collector.backend_service=${SW_AGENT_COLLECTOR_BACKEND_SERVICES:127.0.0.1:11800}
`
   - If extending OAP service, we must restart all service.
  

- HOW ?
  - In this situation, I want to set domain to this variable.The domain has a number of A records
  - modify the method boot() of GRPCChannelManager, check whether it is domain, if so ,then resolve it ,and add all ip addresses to serverList
```
 public void boot() {
        if (Config.Collector.BACKEND_SERVICE.trim().length() == 0) {
            LOGGER.error("Collector server addresses are not set.");
            LOGGER.error("Agent will not uplink any data.");
            return;
        }
        grpcServers = Arrays.asList(Config.Collector.BACKEND_SERVICE.split(","));
        // resolve all domain to ip and add to server List
        connectCheckFuture = Executors.newSingleThreadScheduledExecutor(
            new DefaultNamedThreadFactory("GRPCChannelManager")
        ).scheduleAtFixedRate(
            new RunnableWithExceptionProtection(
                this,
                t -> LOGGER.error("unexpected exception.", t)
            ), 0, Config.Collector.GRPC_CHANNEL_CHECK_INTERVAL, TimeUnit.SECONDS
        );
    }

```
- I know than nginx can help me .But since skywalking already have client loadbalance, I don't want to use nginx, because it could bring some complexity

### PR DESCRIPTION
**Pull request by xuanyu66**
### <Feature description>
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [x] Update the documentation to include this new feature.
- [ ] Tests(including UT, IT, E2E) are added to verify the new feature.

- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #6222.
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java [hunk 1]
```diff
@@ -163,11 +163,14 @@
...
         public static int GET_PROFILE_TASK_INTERVAL = 20;
-
         /**
          * Get agent dynamic config interval
          */
         public static int GET_AGENT_DYNAMIC_CONFIG_INTERVAL = 20;
+        /**
+         * If true, skywalking agent will enable periodically resolving DNS to update receiver service addresses. Otherwise disable
+         */
+        public static boolean DNS_PERIOD_RESOLVE_ACTIVE = false;
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java:173]:
```suggestion
        public static boolean IS_RESOLVE_DNS_PERIODICALLY = false;
```
I think the current naming style is like this.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java [hunk 2]
```diff
@@ -163,11 +163,14 @@
          * Get profile task list interval
          */
         public static int GET_PROFILE_TASK_INTERVAL = 20;
-
         /**
          * Get agent dynamic config interval
          */
         public static int GET_AGENT_DYNAMIC_CONFIG_INTERVAL = 20;
+        /**
+         * If true, skywalking agent will enable periodically resolving DNS to update receiver service addresses. Otherwise disable
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java:171]:
```suggestion
         * If true, skywalking agent will enable periodically resolving DNS to update receiver service addresses. 
```

File: CHANGES.md [hunk 3]
```diff
@@ -37,6 +37,7 @@ Release Notes.
 * Fix RestTemplate plugin recording url tag with wrong port
 * Support collecting logs and forwarding through gRPC.
 * Support config `agent.sample_n_per_3_secs` can be changed in the runtime.
+* Support DNS period resolving mechanism to update backend service.
```
Comment (wu-sheng) [inline CHANGES.md:40]:
```suggestion
* Support DNS periodic resolving mechanism to update backend service addresses.
```

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java [hunk 4]
```diff
@@ -91,6 +97,23 @@ public void shutdown() {
...
+                        .map(InetAddress::getHostAddress)
+                        .map(ip -> String.format("%s:%s", ip, domainAndPort[1]))
+                        .collect(Collectors.toList());
+
+                grpcServers = newGrpcServers;
+            } catch (Throwable t) {
+                LOGGER.error(t, "Failed to resolve {} of backend service.", backendService);
+            }
+        }
+
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java:116]: I think this should be called only when gRPC connection available. `reconnect==true`

File: docs/en/setup/service-agent/java-agent/README.md [hunk 5]
```diff
@@ -95,6 +95,7 @@ property key | Description | Default |
 `collector.grpc_upstream_timeout`|How long grpc client will timeout in sending data to upstream. Unit is second.|`30` seconds|
 `collector.get_profile_task_interval`|Sniffer get profile task list interval.|`20`|
 `collector.get_agent_dynamic_config_interval`|Sniffer get agent dynamic config interval|`20`|
+`collector.dns_period_resolve_active`|If true, skywalking agent will enable periodically resolving DNS to update receiver service addresses. Otherwise disable|`false`|
```
Comment (wu-sheng) [inline docs/en/setup/service-agent/java-agent/README.md:98]:
```suggestion
`collector.dns_period_resolve_active`|If true, skywalking agent will enable periodically resolving DNS to update receiver service addresses. |`false`|
```

Also should follow latest naming suggestion.

Comment (xuanyu66) [reply docs/en/setup/service-agent/java-agent/README.md]: fixed

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java (hunk 4 above)
Comment (xuanyu66) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java]: fixed

File: CHANGES.md (hunk 3 above)
Comment (xuanyu66) [reply CHANGES.md]: fixed

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java (hunk 2 above)
Comment (xuanyu66) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java]: fixed

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java (hunk 1 above)
Comment (xuanyu66) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/Config.java]: fixed

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java [hunk 6]
```diff
@@ -92,6 +98,23 @@ public void shutdown() {
...
+            try {
+                String[] domainAndPort = backendService.split(":");
+
+                List<String> newGrpcServers = Arrays
+                        .stream(InetAddress.getAllByName(domainAndPort[0]))
+                        .map(InetAddress::getHostAddress)
+                        .map(ip -> String.format("%s:%s", ip, domainAndPort[1]))
+                        .collect(Collectors.toList());
+
+                grpcServers = newGrpcServers;
```
Comment (wu-sheng) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java:112]: You are overriding the `grpcServers`, you wouldn't have the chance to resolve DNS again.

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java [hunk 7]
```diff
@@ -92,6 +98,23 @@ public void shutdown() {
     @Override
     public void run() {
         LOGGER.debug("Selected collector grpc service running, reconnect:{}.", reconnect);
+        if (IS_RESOLVE_DNS_PERIODICALLY || reconnect) {
```
Comment (kezhenxu94) [inline apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java:101]:
```suggestion
        if (IS_RESOLVE_DNS_PERIODICALLY && reconnect) {
```

**Review by kezhenxu94 — CHANGES_REQUESTED** [review]

File: apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java (hunk 6 above)
Comment (xuanyu66) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java:112]: I don't get it

Comment (wu-sheng) [reply apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/remote/GRPCChannelManager.java:112]: Oh, sorry, my mistake. You split the config string every time.

**Review by wu-sheng — APPROVED** [review]
LGTM

**Review by kezhenxu94 — APPROVED** [review]

**Pull request merged**